packages/shadycss/src/shadow-parts.js

```diff
@@ -339,9 +363,212 @@ export function onInsertBefore(parentNode, newNode, referenceNode) {
+    // Inject this ::part rule into the array of style rules managed by this
+    // instance. This way, the standard ShadyCSS mechanism for evaluating
+    // per-instance styles will take effect.
+    if (!styleInfo.styleRules.rules.includes(styleNode)) {
+      styleInfo.styleRules.rules.push(styleNode);
+    }
+
+    // We need to explicitly enumerate the active part rules on the styleInfo,
+    // because that is used as part of the per-instance style cache lookup.
+    if (!styleInfo.customPropertyPartRules.includes(styleNode)) {
+      styleInfo.customPropertyPartRules.push(styleNode);
+    }
+  }
+
+  // Augment the array of CSS custom properties consumed by this instance,
+  // because that is also used in the per-instance style cache lookup.
+  for (const property of newProperties) {
+    if (!styleInfo.ownStylePropertyNames.includes(property)) {
+      styleInfo.ownStylePropertyNames.push(property);
+    }
+  }
+
+  // Re-scope this host. This creates a new <style> for this instance, or
+  // re-uses an existing one by checking the cache, and updates "scope" classes
+  // in this host's shadow root.
+  window.ShadyCSS.ScopingShim._applyStyleProperties(host, styleInfo);
+}
+
+/**
+ * A map from "providerScope:receiverScope" compound string key to an array of
+ * ::part rules that were found during template preparation.
+ *
+ * We only populate this map if a ::part rule consumes a custom property and we
+ * native custom properties are not available.
+ *
+ * This map is used to quickly lookup whether a given ::part rule being applied
+ * to an instance of a part node needs special handling for custom properties.
+ *
+ * @type {!Map<string, !Array<!PartRulesMapEntry>>}
+ */
+const partRulesMap = new Map();
+
+/* eslint-disable no-unused-vars */
+/**
+ * An entry in the part rules map. Note this class is just used as a type; it
+ * will be dead-code removed.
+ * @record
+ */
+class PartRulesMapEntry {
+  constructor() {
+    /**
+     * The ShadyCSS StyleNode for this ::part rule.
+     * @type {!StyleNode}
+     */
+    this.styleNode;
+
+    /**
+     * The CSS custom property names consumed by this ::part rule.
+     * @type {!Array<!string>}
+     */
+    this.consumedProperties;
+
+    /**
+     * The parsed part names that this ::part rule targets.
+     * @type {!Array<!string>}
+     */
+    this.partNames;
+  }
+}
+/* eslint-enable no-unused-vars */
+
+/**
+ * Look for any ::part rules in the given AST that consume custom properties,
+ * and load those into the part rules map.
+ *
+ * @param {!string} providerScope Lower-case tag name of the element whose
+ *     template this is.
+ * @param {!StyleNode} styleAst Parsed CSS for this template.
+ */
+export function analyzeTemplatePartRules(providerScope, styleAst) {
+  if (nativeCssVariables || !styleAst.rules) {
+    return;
+  }
+  for (const styleNode of styleAst.rules) {
+    const selector = styleNode['selector'];
+    if (!selector || selector.indexOf('::part') === -1) {
```

Comment:
I was thinking that because this is going to run for every template, I want the cost for users who using part styles at all to be as small as possible. I assumed the regexp would be slower than a regular substring search, although maybe the difference is so negligible that it's better to just re-use the regex, since the tradeoff is that if there *is* a part style, the regex would have sufficed?

Probably needs a benchmark huh? Added a TODO for a benchmark.

(Also I updated `indexOf` to `includes` since it's more readable).